mlfinlab/ensemble/sb_bagging.py

```diff
@@ -0,0 +1,662 @@
+
+    # Draw indices
+    feature_indices = _generate_indices_standard(random_state, bootstrap_features,
+                                                 n_features, max_features)
+    sample_indices = seq_bootstrap(ind_mat, sample_length=max_samples, random_state=random_state)
+
+    return feature_indices, sample_indices
+
+
+def _parallel_build_estimators(n_estimators, ensemble, X, y, ind_mat, sample_weight,
+                               seeds, total_n_estimators, verbose):
+    """Private function used to build a batch of estimators within a job."""
+    # Retrieve settings
+    n_samples, n_features = X.shape
+    max_features = ensemble._max_features
+    max_samples = ensemble._max_samples
+    bootstrap_features = ensemble.bootstrap_features
+    support_sample_weight = has_fit_parameter(ensemble.base_estimator_,
+                                              "sample_weight")
+
+    if not support_sample_weight and sample_weight is not None:
+        raise ValueError("The base estimator doesn't support sample weight")
+
+    # Build estimators
+    estimators = []
+    estimators_features = []
+
+    for i in range(n_estimators):
+        if verbose > 1:
+            print("Building estimator %d of %d for this parallel run "
+                  "(total %d)..." % (i + 1, n_estimators, total_n_estimators))
+
+        random_state = np.random.RandomState(seeds[i])
+        estimator = ensemble._make_estimator(append=False,
+                                             random_state=random_state)
+
+        # Draw random feature, sample indices
+        features, indices = _generate_bagging_indices(random_state,
+                                                      bootstrap_features,
+                                                      n_features,
+                                                      max_features,
+                                                      max_samples,
+                                                      ind_mat)
+
+        # Draw samples, using sample weights, and then fit
+        if support_sample_weight:
+            if sample_weight is None:
+                curr_sample_weight = np.ones((n_samples,))
+            else:
+                curr_sample_weight = sample_weight.copy()
+
+            sample_counts = np.bincount(indices, minlength=n_samples)
+            curr_sample_weight *= sample_counts
+
+            estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)
+
+        else:
+            estimator.fit((X[indices])[:, features], y[indices])
+
+        estimators.append(estimator)
+        estimators_features.append(features)
+
+    return estimators, estimators_features
+
+
+class SequentiallyBootstrappedBaseBagging(BaseBagging, metaclass=ABCMeta):
+    """
+    Base class for Sequentially Bootstrapped Classifier and Regressor, extension of sklearn's BaseBagging
+    """
+
+    @abstractmethod
+    def __init__(self,
+                 triple_barrier_events,
```

Comment:
I think we should use a different name for this parameter as users might use a different way to generate important events other than the Triple Barrier method. This is one of the reasons I used the name `info_sets` in my naming on the Cross Validation chapter. Don't know if my name is the best, but I think sticking to to `triple_barrier_events` is limiting. Maybe `events_info_set` -- as in "the information set that was used to generate a particular event" What do you think ?

Reply:
What about events_end_times?

Reply:
Well, `events_end_times` is not totally accurate as the index actually indicates the start time of the event, so the entire data structure has both the start and end of the event. Other suggestions ?

Reply:
What about labels_info_sets or samples_info_sets?

Reply:
`samples_info_sets` sounds good to me

Reply:
perfect